src/overlays/actors/ovl_En_Niw/z_en_niw.c

```diff
@@ -21,61 +42,1136 @@ const ActorInit En_Niw_InitVars = {
+void EnNiw_ParticleSpawn(EnNiw* this, Vec3f* pos, Vec3f* vel, Vec3f* accel, f32 scale) {
+    s16 i;
+    EnNiwParticle* particle = this->particle;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Niw/func_80AB7290.s")
+    for (i = 0; i < ARRAY_COUNT(this->particle); i++, particle++) {
+        if (particle->type == 0) {
+            particle->type = 1;
+            particle->pos = *pos;
+            particle->vel = *vel;
+            particle->accel = *accel;
+            particle->timer = 0;
+            particle->scale = scale / 1000.0f;
+            particle->lifespan = (s16)Math_Rand_ZeroFloat(20.0f) + 40;
+            particle->unk_2A = Math_Rand_ZeroFloat(1000.0f);
+            return;
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Niw/func_80AB7328.s")
+void EnNiw_ParticleUpdate(EnNiw* this, GlobalContext* globalCtx) {
+    s16 i;
+    EnNiwParticle* particle = this->particle;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Niw/func_80AB7420.s")
+    for (i = 0; i < ARRAY_COUNT(this->particle); i++, particle++) {
+        if (particle->type != 0) {
+            particle->timer++;
+            particle->pos.x += particle->vel.x;
+            particle->pos.y += particle->vel.y;
+            particle->pos.z += particle->vel.z;
+            particle->vel.x += particle->accel.x;
+            particle->vel.y += particle->accel.y;
+            particle->vel.z += particle->accel.z;
+            if (particle->type == 1) {
+                particle->unk_2A++;
+                Math_SmoothScaleMaxF(&particle->vel.x, 0.0f, 1.0f, 0.05f);
+                Math_SmoothScaleMaxF(&particle->vel.z, 0.0f, 1.0f, 0.05f);
+                if (particle->vel.y < -0.5f) {
+                    particle->vel.y = -0.5f;
+                }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Niw/func_80AB747C.s")
+                particle->unk_30 = Math_Sins(particle->unk_2A * 0xBB8) * M_PI * 0.2f;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Niw/EnNiw_Update.s")
+                if (particle->lifespan < particle->timer) {
+                    particle->type = 0;
+                }
+            }
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Niw/func_80AB7F60.s")
+void EnNiw_ParticleDraw(EnNiw* this, GlobalContext* globalCtx) {
+    u8 flag = 0;
+    s16 i;
+    s32 pad;
+    GraphicsContext* gfxCtx = globalCtx->state.gfxCtx;
+    EnNiwParticle* particle = this->particle;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Niw/EnNiw_Draw.s")
+    OPEN_DISPS(gfxCtx, "../z_en_niw.c", 1897);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Niw/func_80AB8168.s")
+    func_80093D84(globalCtx->state.gfxCtx);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Niw/func_80AB8258.s")
+    for (i = 0; i < ARRAY_COUNT(this->particle); i++, particle++) {
+        if (particle->type == 1) {
```

Comment:
Same thing as the loop above. I don't think you need to manually create a particle iterator temp and you can just use this->particle[i], hopefully

Reply:
Unfortunately this does not match